### PR TITLE
Fixed #28663 -- Add a check for likely incorrectly migrated django.urls.path() routes.

### PR DESCRIPTION
**Pull request by lamby**
https://code.djangoproject.com/ticket/28663

### CODE REVIEW

File: django/urls/resolvers.py [hunk 1]
```diff
@@ -256,6 +256,16 @@ def match(self, path):
...
 
     def check(self):
+        if self._route.startswith('^') or self._route.endswith('$'):
+            warning = Warning(
+                "Your URL pattern {} has a route beginning with a '^' or ends "
+                "with a '$'. This was likely an accident after migrating to "
+                "django.urls.path for Django 2.x".format(
+                    self.describe()
+                ),
+                id="urls.W004",
```
Comment (felixxm) [inline django/urls/resolvers.py]: Please use single quotes.

File: django/urls/resolvers.py [hunk 2]
```diff
@@ -256,6 +256,16 @@ def match(self, path):
...
+            )
+            return [warning]
```
Comment (felixxm) [inline django/urls/resolvers.py]: I think we should return a list of all warnings, e.g.:
```python
def check(self):
    warnings = self._check_pattern_startswith_slash()
    if self._route.startswith('^') or self._route.endswith('$'):
        warnings.append(Warning(...))
    return warnings
```

File: django/urls/resolvers.py (hunk 1 above)
Comment (felixxm) [inline django/urls/resolvers.py]: Please add `urls.W004` to the `ref/checks.txt`.

File: django/urls/resolvers.py [hunk 3]
```diff
@@ -256,6 +256,16 @@ def match(self, path):
         return None
 
     def check(self):
+        if self._route.startswith('^') or self._route.endswith('$'):
+            warning = Warning(
+                "Your URL pattern {} has a route beginning with a '^' or ends "
+                "with a '$'. This was likely an accident after migrating to "
+                "django.urls.path for Django 2.x".format(
```
Comment (felixxm) [inline django/urls/resolvers.py]: Dot is missing _"... for Django 2.x."_.

**Comment by lamby** [comment]
Thanks for the review. PR updated to address all issues.

File: django/urls/resolvers.py [hunk 4]
```diff
@@ -256,7 +256,17 @@ def match(self, path):
         return None
 
     def check(self):
-        return self._check_pattern_startswith_slash()
+        warnings = self._check_pattern_startswith_slash()
+        if self._route.startswith('^') or self._route.endswith('$'):
```
Comment (orf) [inline django/urls/resolvers.py]: Perhaps adding a check for the route containing `(?P<` could be a good idea as well?

Comment (lamby) [reply django/urls/resolvers.py]: Ooh, nice one...

**Comment by lamby** [comment]
Updated to match all review.

File: docs/releases/2.1.txt [hunk 5]
```diff
@@ -183,7 +183,10 @@ Tests
 URLs
 ~~~~
 
-* ...
+* Added a system check warning if an URL pattern passed to
```
Comment (timgraham) [inline docs/releases/2.1.txt]: No need for a release note -- anyway, I think this would be a candidate for 2.0.

**Comment by timgraham** [comment]
Chris, do you have any thoughts about whether this check should be temporary or permanent as I suggested in my earlier comment? I suppose it's unlikely that URLs would be designed to trigger this check in a false positive fashion but many of the "helpful" upgrade checks we've tried in the past have trigger false positives for some users.

**Comment by lamby** [comment]
@timgraham  wrote:

> whether this check should be temporary or permanent as I suggested in my earlier comment?

Sorry, I did not see your previous comments until now.

I feel this should be "permanent" in the sense of lasting a fair amount into the 2.x cycle - I can imagine a bunch of 1.x projects being upgraded directly to (say) 2.2 or 2.3 in a couple of years and will thus hit this. In other words, whether ``url`` goes away is somewhat irrelevant - users can still make the mistake when migrating.

(Is that that the only thing blocking this PR, ooi? Let me know if not and I will address it.)

**Pull request merged**